lib/elixir/lib/file.ex

```diff
@@ -931,6 +931,11 @@ defmodule File do
   option when opening the file and then all other functions from
   `IO` are available, since they work directly with Unicode data.
 
+  `modes_or_function` can either be a list of modes or a function. If it's a
+  list, it's considered to be a list of modes (that are documented below). If
+  it's a function, then it's equivalent to calling `open(path, [],
+  modes_or_function)`. See the documentation for `open/3` for more information.
```

Comment:
That's a good point, but this is a consequence of how the code works today: if `modes_or_function` is a list of modes, we call to `:file`, but if it's a function, we call to `open/3` which calls to `open/2` with a list of modes and so on.

I see two options: the first is we keep this code and rework the documentation. The other is we implement `File.open/2` as:

```elixir
def open(path, modes) when is_list(modes), do: open(path, modes, &(&1))
def open(path, function) when is_function(function, 1), do: open(path, [], function)
```

I believe this shouldn't bee bad for performance as opening a file is likely to take most of the time here. Wdyt?